Fix citizens abandoning their cars in strange places

WildCard-25 opened the pull request:
This patch forces citizens to use their cars when going home if their car isn't near their home. It also allows them to use public transport to get to their car.
It also fixes citizens that have an abandoned car and need to use a car to go somewhere from becoming confused and staying home (if they became confused when not at home, game teleports them home, possibly abandoning their car).
Just despawning their car in these instances doesn't work as they never spawn a new car afterwards, so I've moved their parked car closer instead, except for cases where the parked car is already near their home.

An example of what this patch does:-
Before patch: Citizen drives to work, takes public transport to shops, then takes public transport home, leaving their car at work.
After patch: Citizen drives to work, takes public transport to shops, then takes public transport back to car, then drives home.
Not yet implemented: Citizen drives to work, drives to shops, then drives home.

WildCard-25 commented:
Implemented: Citizen drives to work, drives to shops, then drives home.
Would also allow: drive to public transport->work/shop, as travelling home will go via public transport to car, then home.

Don't know how this will interact with commit of 'Citizens are forced to return their car back home' on 1.10.6 branch from 17hrs ago yet :-)

WildCard-25 commented:
After a quick look over 'Citizens are forced to return their car back home', it looks like it still has some of the bugs my patch fixes.

iirc the main reason I moved the 'if (parkedVehicleId !=0)' before the 'if (extInstance.pathMode == ExtPathMode.CalculatingWalkingPathToParkedCar etc) block, and the pathmode change code that then triggers the pathmode block. When a citizen has carUsagePolicy.Forced that doesn't prevent him from just walking home if his car is further away from home than he is. It also stops him from using public transport to get to his car first, then drive it home. If his car and home were both too far to walk to, then path finding would fail and the citizen would teleport home, abandoning his car.

My first patch fixes the abandoned car scenarios. Before the fix cars never despawned like they are supposed to with a hard path failure, fixed that and then they never re-spawned, so I opted for moving them instead. This also stops them from being confused at home all the time if the only way to get to places (like work) is by using their car, but their car is on the other side of the city due to them teleporting home at some point in their past.
My second patch fixes the walking home instead of using car and public transport to car, then car to home scenarios.
I still need to test your code to be 100% sure, but I'm 95% sure it doesn't fix this yet :-)

VictorPhilipp commented:
First of all thank you for your contribution. Don't start to test my code yet, version 1.10.6 is still under development and the new logic not yet fully integrated/functional. You are right, I used _carUsagePolicy_ incorrectly in the latest commit.

The main idea is to remember the last mode of transport that the citizen used to reach their current location. If the citizen is not at home, the last mode of transport (e.g. driving a car) will be used again to move to the next destination. On the other hand, if the citizen is at home they may use whatever mode of transport they like (public transport, car, bike, etc.).

Regarding your changes in _AdvancedParkingManger_:
I like your idea of moving the parked car whenever it is out of range. However, I think this should only be used as a fallback method (e.g. in case the citizen uses teleportation magic). I saw that you move the car after path-finding has failed. I think it would be better if this is done before path-finding. We have to make sure that path-finding does not retry endlessly: Moving the car to a closer location in no way guarantees that it becomes reachable by walking. The logic also must account for tourists: They are effectively "homeless" and moving their car closer to home will not work as it is implemented.

Regarding _CustomCitizenAI_:
In your version, if _pathMode == CalculatingWalkingPathToParkedCar_, the variable _carUsageMode_ is set to _Allowed_ where it must be _Forbidden_ (because we must calculate a dedicated walking path from source to the citizen's parked car).

You are right in that _carUsageMode == Forced_ prevents citizens from using public transport. This is intended because path-finding is almost always performed with _carUsageMode == Allowed_ and _pathMode == None_ first (except if we force the citizen to return their car back home). If this first path contains a car section we split the path into up to three sections:
1. Walk to parked car (path modes *WalkingPathToParkedCar)
2. Drive to target building (*CarPathToTarget) OR a predetermined parking space (*CarPathToKnownParkPos)
3. Walk to target building (if we drove to a predetermined parking space) (*WalkingPathToTarget)

I will update the 1.10.6 branch within the next few days and will report back when I finished the next iteration.

WildCard-25 commented:
Hi Victor,
Thanks for your feedback, however your interpretation of how my code interacts with the existing code is in error. I have actually tested this code for both residents and tourists.

Regarding _AdvancedParkingManger_:
This code is only triggered after an initial path find failure, then if the parked vehicle is to far from the citizen AND to far from the citizens home it is moved closer to the citizen and path finding soft fails so that it may be recalculated from scratch.
Once the vehicle has been moved **once** and path finding fails again, it is already close to the citizen and won't be moved again. This second failure results in a hard fail and path finding stops.
Checking if _homeId == 0_ could be used to bypass the parkedToHome magnitude calculation for tourists, but for tourists this calculation is close to maxInt so, for tourists, this if block is always executed.
If the car can't be moved, or any of the other previous conditions aren't meet, then this results in a hard failure and path finding stops.

Regarding _CustomCitizenAI_:
if _pathMode == CalculatingWalkingPathToParkedCar_ then the following if block (ie. _if (extInstance.pathMode == ExtPathMode.CalculatingWalkingPathToParkedCar || ...)_) is triggered and sets _carUsageMode_ to _Forbidden_
With my code, if the residents (not tourist) is heading home and their car is not near their home, then:
1. They first go to their car, using public transport if required.
2. Drive home OR a predetermined parking space near home.
3. Walk home (if we drove to a predetermined parking space)

I have tested this code for both residents and tourists. Testing code has been removed, as I'm sure you don't want all the code that was printing all the test variables before each if statement left in.

You'll also notice that I also completed testing on the second change to _CustomCitizenAI_ that will sometimes make a citizen use their car when travelling from one location to another location other than home. This code also works as expected for both residents and tourists.
Conditions: Citizens car must **not** be parked near home, **and** their parked car must be nearby, **and** their parked car must **not** be near new location. Works for residents and tourists.

Cheers Daniel.
p.s already tried the new code but figured it was incomplete when I got the 'Index Out Of Range' exceptions.

WildCard-25 commented:
Lots of interesting changes in alpha-2 commit, will have a look and test later (time permitting) :-)

In the meantime I tried to find a simple example of an abandoned car to test with (I deleted my old test town over a month ago). Here is a short [video of a student](https://drive.google.com/open?id=1c0DxnAP0_cH6_e7g8E0NGhJIAbH6Q3jf) who has an abandoned car, we can assume the student either used public transport or walked to get to school. This shows 1.10.5 behaviour, behaviour with the 'Citizens are forced to return their car back home' patch, and my own patch.

I'll try and find examples of the other behaviours I fixed as well and retest them against the new alpha-2 code.

VictorPhilipp commented:
Thanks for the clarification. You are right, the car relocation code inside _AdvancedParkingManager_ works as you describe it. As you can see I integrated a slightly modified version of your code. At the same time I also solved two problems related to spawning new parked cars:
1. Previously, the method _TrySpawnParkedPassengerCar_ did not search for building parking spaces when a road side parking space was found. In areas where road-side parking is impossible this caused the parked vehicle to be spawned very far away from the starting position. This in turn caused your car relocation code to fire very often.
2. The code in _OnCitizenPathFindSuccess_ that tries to determine the current position vector based on the current building was incorrect. This often caused parked cars to be spawned near the starting building even if the citizen was already en route.

You are also right regarding _CustomCitizenAI_: I missed that in your version actually both blocks
1. _if (parkedVehicleId != 0 && extInstance.pathMode != ExtCitizenInstance.ExtPathMode.CalculatingWalkingPathToTarget)_ and
2. _if (extInstance.pathMode == ExtPathMode.CalculatingWalkingPathToParkedCar ...)_
get executed since you removed the _else_ statement. The first block sets _carUsageMode_ to _Allowed_ but then the next block sets it to _Forbidden_. :-)

In the alpha-2 code I restructured the _ExtPathFind_ code a bit. Storing and retrieving the mode of transport the citizen used to leave home should now be stored and retrieved correctly.

VictorPhilipp commented:
Regarding the index out of bounds exceptions: I added a new entry to the _Switches_ array in the _Debug_ class. Since the entry is not present during deserialization of the global config file this causes said exceptions.

WildCard-25 commented:
Thank you for the info about the index out of bounds exception.
Finally had some time to look and test a bit :-)

So, basically, the changes from '[Citizens not using their car to get home](https://github.com/VictorPhilipp/Cities-Skylines-Traffic-Manager-President-Edition/pull/108/commits/72b91755ae1ec46c6061876d931934d2fdc2aea4)' are now integrated.
Most of the changes from '[Citizens being confused due to abandoned car](https://github.com/VictorPhilipp/Cities-Skylines-Traffic-Manager-President-Edition/pull/108/commits/1924a833f4d14bc95b55cbfc15107679bd54baa3)' and much cleaner than my 'wall-o-code', although, in my defense, I was trying to match the existing style. ;-)

The missing 2 bits being:-
1. the _parkedToHome_ check (not needed if only the _pathmode == CalculatingWalkingPathToParkedCar_ condition is allowed through the if statement as it's already been checked earlier on in the code path).
2. the _pathMode == None_ condition being allowed into the if block.

So, why are these bits needed? (Documentation, the ever present coders bane.)
Say a cim is at home and wants to go to work, but their car has been abandoned at their work, it is too far to walk and there is no public transport (this is actually the test case I originally used). Path finding will fail and enter _OnCitizenPathFindFailure_ with a _extInstance.pathMode_ of _None_. With this code, their car will be moved and they can try again. Without this code they become _Confused_ and stay at home.
Of course, _eventually_, they will go somewhere else and when they try to go home this code will finally kick in and move their car allowing them to drive home and then drive to work.

I haven't had time to see how well they use their cars when travelling from location to location yet, compared to how they work with my '[Citizens should use their cars from location to location](https://github.com/VictorPhilipp/Cities-Skylines-Traffic-Manager-President-Edition/pull/108/commits/63d4e7f3ba5ad0b9717c58ee7db96b92cf86c1ba)' patch yet.

And lastly, unfortunately, the U-turn bug I noticed is still there. Will provide more info after I check a couple of things.

VictorPhilipp commented:
1.
Im with you - I don't believe that we need the _parkedToHome_ check in _OnCitizenPathFindFailure_. since this is already handled by _CustomCitizenAI.ExtStartPathFind_ . There, _pathMode_ switches to _CalculatingWalkingPathToParkedCar_ if
- the citizen wants to go home and
- their parked car is located too far away from home.
In _AdvancedParkingManager.OnCitizenPathFindFailure_ the only situation that remains to be handled is when the parked car is unreachable.

2.
The car relocation code ~~does not~~ should not need to get executed if _pathMode = None_:
_pathMode = None_ denotes the first stage of path-finding where under normal circumstances* we just do a vanilla path calculation (source to target building). The resulting path ~~is~~ should be unconstraint, that means pocket car spawning ~~is~~ should be allowed and parking space search is not performed.
Thus if path-finding fails while _pathMode = None_, there should be really no chance to find a path to the target building.

While writing this part I just discovered that pocket car spawning is indeed currently not always allowed during the first stage. **This is a bug**:
Assume _pathMode = None_ and _ExtStartPathFind_. is called.  
- If a parked vehicle is present (_parkedVehicleId ≠ 0_), _vehicleInfo_ is overwritten (_CustomCitizenAI_, line 114).
- We determine the _extVehicleType_ of _vehicleInfo_ (lines 217-251). Assume _extVehicleType = PassengerCar_.
- Assuming that everything went fine and that _pathMode_ is still _None_: The path position of the parked vehicle is fetched (_parkedVehiclePathPos_, output parameter in line 346).
- _parkedVehiclePathPos_ is passed to path-finding (line 429).
- _parkedVehiclePathPos_ (via _m_position11_) finally arrives in _CustomPathFind_, fields __vehicleLane_ and __vehicleOffset_ (lines 379-385).
- And then the bug occurs (note that path-finding works backwards, from target to source - previous lane = nearer to target, next lane = nearer to source): The _else_ block starting at line 979 is executed when the previous lane was a vehicle lane (citizen drives car to target). Lines 992-995 control where switching from vehicle to pedestrian lane may happen --> only at the position of the parked vehicle --> this is wrong when Parking AI is active and _pathMode = None_ since lane switching should be allowed everywhere.

Regarding code style, and the history of this mod:

Well, I took iMarbot's code and then started with some trial & error coding to understand how the game works. This code then went "into production". I really would love to have some unit/integration tests at hand but this would require a working testing framework and/or a mod reloading feature in the game. Without that of course, restructuring code, improving readability, optimizations, etc. produces new errors and solving errors is time consuming due to the required game restart(s).

Somewhere in 2017 I started to decouple _Custom*AI_ classes from _*Manager_ classes, and _*Manager_ classes from game code: While the project _TMPE.GenericGameBridge_ provides an API, _TMPE.CitiesGameBridge_ implements it for production and (the yet empty project) _TMPE.TestGameBridge_ would then implement the API for testing purposes. Whenever possible (and time-dependent) I abstract everything game dependent away and add new API methods to _TMPE.GenericGameBridge_ (as done with the _IVehicleService.ProcessParkedVehicle_ method which I added for your case).

Somewhere in the future I see a fully decoupled TMPE which then can be used in Sim City :-P

WildCard-25 commented:
Ok, I think I see where our interpretations differ now. :-)

I was assuming that not spawning a pocket car when a citizen already had a parked car and parking AI (prohibitPocketCars) was turned on was deliberate. So by following the [decision tree](https://www.viathinksoft.de/tmpe/wiki/images/a/a6/Parking-ai.png), from the wiki, after _considering all available modes of transport: find best path from source to target_ and no path is found it results in a hard fail. Instead of the hard fail, I then moved their parked vehicle and let them try again.

What you want to do is include the use of pocket cars during the initial _considering all available modes of transport: find best path from source to target_ block. And then as some point move their parked car to where the pocket car would have been spawned along their path?

I noticed a lot of cims using their cars when they shouldn't be (ie. already parked near their house, yet using their car to go home). To fix, change line 127 of _CustomCitizenAI_ from
`if ((extCitizen.lastTransportMode | ExtCitizen.ExtTransportMode.Car) != ExtCitizen.ExtTransportMode.None) {`
to
`if ((extCitizen.lastTransportMode & ExtCitizen.ExtTransportMode.Car) != ExtCitizen.ExtTransportMode.None) {`

And the U-turn bug.
![u-turn bug](https://user-images.githubusercontent.com/1796574/35921396-34449e6e-0c66-11e8-9548-bd993e27e178.jpg)
This happens whether or not _Vehicle may do u-turns at junctions_ is active.

As to coding style, I was already using a similar style to Steve long before he brought out his first book (see 'Tcl/Tk Engineering Manual' and 'Tcl Style Guide'). Yes, I'm old :-)
